homeassistant/components/smartthings/smartapp.py

```diff
@@ -0,0 +1,263 @@
+    manager = SmartAppManager(
+        webhook.async_generate_path(config[CONF_WEBHOOK_ID]),
+        dispatcher=dispatcher)
+    manager.connect_install(functools.partial(smartapp_install, hass))
+    manager.connect_uninstall(functools.partial(smartapp_uninstall, hass))
+
+    webhook.async_register(hass, DOMAIN, 'SmartApp',
+                           config[CONF_WEBHOOK_ID], smartapp_webhook)
+
+    hass.data[DOMAIN] = {
+        DATA_MANAGER: manager,
+        CONF_INSTANCE_ID: config[CONF_INSTANCE_ID],
+        DATA_BROKERS: {},
+        CONF_WEBHOOK_ID: config[CONF_WEBHOOK_ID]
+    }
+
+
+async def smartapp_install(hass: HomeAssistantType, req, resp, app):
+    """
+    Handle when a SmartApp is installed by the user into a location.
+
+    Setup subscriptions using the access token SmartThings provided in the
+    event. An explicit subscription is required for each 'capability' in order
+    to receive the related attribute updates.  Finally, create a config entry
+    representing the installation.
+    """
+    from pysmartthings import SmartThings, Subscription, SourceType
+
+    api = SmartThings(async_get_clientsession(hass), req.auth_token)
+
+    async def create_subscription(target):
+        sub = Subscription()
+        sub.installed_app_id = req.installed_app_id
+        sub.location_id = req.location_id
+        sub.source_type = SourceType.CAPABILITY
+        sub.capability = target
+        try:
+            await api.create_subscription(sub)
+            _LOGGER.debug("Created subscription for '%s' under app '%s'.",
+                          target, req.installed_app_id)
+        except Exception:  # pylint:disable=broad-except
+            _LOGGER.exception("Failed to create subscription for '%s' under "
+                              "app '%s'", target, req.installed_app_id)
+
+    tasks = [create_subscription(c) for c in SUPPORTED_CAPABILITIES]
+    await asyncio.gather(*tasks)
+    _LOGGER.debug("SmartApp '%s' under parent app '%s' was installed.",
+                  req.installed_app_id, app.app_id)
+
+    # Add as job not needed because the current coroutine was invoked
+    # from the dispatcher and is not being awaited.
+    await hass.config_entries.flow.async_init(
+        DOMAIN, context={'source': 'install'},
+        data={
+            CONF_APP_ID: app.app_id,
+            CONF_INSTALLED_APP_ID: req.installed_app_id,
+            CONF_LOCATION_ID: req.location_id
+        })
+
+
+async def smartapp_uninstall(hass: HomeAssistantType, req, resp, app):
+    """
+    Handle when a SmartApp is removed from a location by the user.
+
+    Find and delete the config entry representing the integration.
+    """
+    entry = next((entry for entry in hass.config_entries.async_entries(DOMAIN)
+                  if entry.data.get(CONF_INSTALLED_APP_ID) ==
+                  req.installed_app_id),
+                 None)
+    if entry:
+        _LOGGER.debug("SmartApp '%s' under parent app '%s' was removed.",
```

Comment:
We don't end logging messages with period.

Reply:
Removed all periods.